File: network_frame_cpp/src/receiver/receiver.cpp
#include "receiver/receiver.h"
#include <fstream>
#include <iostream>
#include <iomanip>

namespace receiver {

ParsedPacket Receiver::decapsulate(const std::vector<uint8_t>& data) {
    ParsedPacket result;
    
    std::cout << "=== Starting Decapsulation ===" << std::endl;
    std::cout << "Total bytes received: " << data.size() << std::endl << std::endl;
    
    // 1. 解析以太网帧
    std::cout << "--- Data Link Layer (Ethernet II) ---" << std::endl;
    auto ethFrame = datalink::EthernetFrame::decode(data);
    
    std::cout << "  Dest MAC: " << datalink::EthernetFrame::macToString(ethFrame.getHeader().dstMAC) << std::endl;
    std::cout << "  Source MAC: " << datalink::EthernetFrame::macToString(ethFrame.getHeader().srcMAC) << std::endl;
    std::cout << "  EtherType: 0x" << std::hex << std::setfill('0') << std::setw(4) 
              << ethFrame.getHeader().etherType << std::dec << std::endl;
    
    if (!ethFrame.isIPv4()) {
        throw std::runtime_error("EtherType is not IPv4 (0x0800)");
    }
    std::cout << "  EtherType verified: IPv4" << std::endl;
    
    result.ethernetFrame = std::make_unique<datalink::EthernetFrame>(ethFrame);
    
    // 2. 解析IPv4数据报
    std::cout << "\n--- Network Layer (IPv4) ---" << std::endl;
    auto ipPacket = network::IPv4Packet::decode(ethFrame.getPayload());
    
    auto ipHeader = ipPacket.getHeader();
    std::cout << "  Version: " << static_cast<int>(ipHeader.version) << std::endl;
    std::cout << "  Header Length: " << (ipHeader.ihl * 4) << " bytes" << std::endl;
    std::cout << "  Total Length: " << ipHeader.totalLength << " bytes" << std::endl;
    std::cout << "  TTL: " << static_cast<int>(ipHeader.ttl) << std::endl;
    std::cout << "  Protocol: " << static_cast<int>(ipHeader.protocol) << std::endl;
    std::cout << "  Source IP: " << network::IPv4Packet::ipToString(ipHeader.srcIP) << std::endl;
    std::cout << "  Dest IP: " << network::IPv4Packet::ipToString(ipHeader.dstIP) << std::endl;
    
    if (!ipPacket.isUDP()) {
        throw std::runtime_error("Protocol is not UDP (17)");
    }
    std::cout << "  Protocol verified: UDP (17)" << std::endl;
    
    result.ipv4Packet = std::make_unique<network::IPv4Packet>(ipPacket);
    
    // 3. 解析UDP数据报
    std::cout << "\n--- Transport Layer (UDP) ---" << std::endl;
    auto udpDatagram = transport::UDPDatagram::decode(ipPacket.getPayload());
    
    auto udpHeader = udpDatagram.getHeader();
    std::cout << "  Source Port: " << udpHeader.srcPort << std::endl;
    std::cout << "  Dest Port: " << udpHeader.dstPort << std::endl;
    std::cout << "  UDP Length: " << udpHeader.length << " bytes" << std::endl;
    std::cout << "  Checksum: 0x" << std::hex << std::setfill('0') << std::setw(4) 
              << udpHeader.checksum << std::dec << std::endl;
    
    result.udpDatagram = std::make_unique<transport::UDPDatagram>(udpDatagram);
    
    // 4. 还原应用层数据
    std::cout << "\n--- Application Layer ---" << std::endl;
    auto appData = application::Data(udpDatagram.getPayload());
    
    std::cout << "  Data: " << appData.getPayloadString() << std::endl;
    std::cout << "  Size: " << appData.size() << " bytes" << std::endl;
    
    result.applicationData = std::make_unique<application::Data>(appData);
    
    std::cout << "\n=== Decapsulation Complete ===" << std::endl;
    
    return result;
}

ParsedPacket Receiver::decapsulateFromFile(const std::string& filename) {
    std::cout << "Reading file: " << filename << std::endl << std::endl;
    
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Failed to open file for reading: " + filename);
    }
    
    std::vector<uint8_t> data((std::istreambuf_iterator<char>(file)),
                               std::istreambuf_iterator<char>());
    file.close();
    
    return decapsulate(data);
}

std::string Receiver::getApplicationData(const std::string& filename) {
    auto parsed = decapsulateFromFile(filename);
    return parsed.applicationData->getPayloadString();
}

} // namespace receiver

File: network_frame/pkg/receiver/receiver.go
package receiver

import (
	"errors"
	"fmt"
	"os"

	"network_frame/pkg/application"
	"network_frame/pkg/datalink"
	"network_frame/pkg/network"
	"network_frame/pkg/transport"
)

// ParsedPacket 解析后的数据包结构
type ParsedPacket struct {
	EthernetFrame   *datalink.EthernetFrame
	IPv4Packet      *network.IPv4Packet
	UDPDatagram     *transport.UDPDatagram
	ApplicationData *application.Data
}

// Receiver 解封装模块
type Receiver struct{}

// NewReceiver 创建新的接收器
func NewReceiver() *Receiver {
	return &Receiver{}
}

// Decapsulate 解封装数据
func (r *Receiver) Decapsulate(data []byte) (*ParsedPacket, error) {
	result := &ParsedPacket{}

	fmt.Println("=== Starting Decapsulation ===")
	fmt.Printf("Total bytes received: %d\n\n", len(data))

	// 解析以太网帧
	fmt.Println("--- Data Link Layer (Ethernet II) ---")
	ethFrame, err := datalink.DecodeEthernet(data)
	if err != nil {
		return nil, fmt.Errorf("failed to decode Ethernet frame: %w", err)
	}
	result.EthernetFrame = ethFrame

	fmt.Printf("  Dest MAC: %s\n", ethFrame.Header.DstMAC)
	fmt.Printf("  Source MAC: %s\n", ethFrame.Header.SrcMAC)
	fmt.Printf("  EtherType: 0x%04X\n", ethFrame.Header.EtherType)

	if !ethFrame.IsIPv4() {
		return nil, errors.New("EtherType is not IPv4 (0x0800)")
	}
	fmt.Println("  EtherType verified: IPv4")

	// 解析IPv4数据报
	fmt.Println("\n--- Network Layer (IPv4) ---")
	ipPacket, err := network.DecodeIPv4(ethFrame.GetPayload())
	if err != nil {
		return nil, fmt.Errorf("failed to decode IPv4 packet: %w", err)
	}
	result.IPv4Packet = ipPacket

	fmt.Printf("  Version: %d\n", ipPacket.Header.Version)
	fmt.Printf("  Header Length: %d bytes\n", ipPacket.Header.IHL*4)
	fmt.Printf("  Total Length: %d bytes\n", ipPacket.Header.TotalLength)
	fmt.Printf("  TTL: %d\n", ipPacket.Header.TTL)
	fmt.Printf("  Protocol: %d\n", ipPacket.Header.Protocol)
	fmt.Printf("  Source IP: %s\n", ipPacket.Header.SrcIP)
	fmt.Printf("  Dest IP: %s\n", ipPacket.Header.DstIP)

	if !ipPacket.IsUDP() {
		return nil, fmt.Errorf("protocol is not UDP (17), got: %d", ipPacket.Header.Protocol)
	}
	fmt.Println("  Protocol verified: UDP (17)")

	// 解析UDP数据报
	fmt.Println("\n--- Transport Layer (UDP) ---")
	udpDatagram, err := transport.DecodeUDP(ipPacket.GetPayload())
	if err != nil {
		return nil, fmt.Errorf("failed to decode UDP datagram: %w", err)
	}
	result.UDPDatagram = udpDatagram

	fmt.Printf("  Source Port: %d\n", udpDatagram.Header.SrcPort)
	fmt.Printf("  Dest Port: %d\n", udpDatagram.Header.DstPort)
	fmt.Printf("  UDP Length: %d bytes\n", udpDatagram.Header.Length)
	fmt.Printf("  Checksum: 0x%04X\n", udpDatagram.Header.Checksum)

	// 还原应用层数据
	fmt.Println("\n--- Application Layer ---")
	appData := application.NewDataFromBytes(udpDatagram.GetPayload())
	result.ApplicationData = appData

	fmt.Printf("  Data: %s\n", appData.GetPayloadString())
	fmt.Printf("  Size: %d bytes\n", appData.Size())

	fmt.Println("\n=== Decapsulation Complete ===")

	return result, nil
}

// DecapsulateFromFile 从文件读取并解封装数据
func (r *Receiver) DecapsulateFromFile(filename string) (*ParsedPacket, error) {
	fmt.Printf("Reading file: %s\n\n", filename)

	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	return r.Decapsulate(data)
}

// GetApplicationData 从文件中提取应用层数据字符串
func (r *Receiver) GetApplicationData(filename string) (string, error) {
	parsed, err := r.DecapsulateFromFile(filename)
	if err != nil {
		return "", err
	}
	return parsed.ApplicationData.GetPayloadString(), nil
}
